pyccl/base.py

```diff
@@ -0,0 +1,47 @@
+import sys
+from collections import OrderedDict
+import numpy as np
+
+
+def _to_hashable(obj):
+    """Make unhashable objects hashable in a consistent manner."""
+
+    if isinstance(obj, (int, float, str)):
+        # Strings and Numbers are hashed directly.
+        return obj
+
+    elif hasattr(obj, "__iter__"):
+        # Encapsulate all the iterables to quickly discard as needed.
+
+        if isinstance(obj, np.ndarray):
+            # Numpy arrays: Convert the data buffer to a byte string.
+            return obj.tobytes()
+
+        elif isinstance(obj, dict):
+            # Dictionaries: Build a tuple from key-value pairs,
+            # where all values are converted to hashables.
+            out = dict.fromkeys(obj)
+            for key, value in obj.items():
+                out[key] = _to_hashable(value)
+            # Sort unordered dictionaries for hash consistency.
+            if isinstance(obj, OrderedDict):
+                return tuple(obj.items())
+            return tuple(sorted(obj.items()))
+
+        else:
+            # Iterables: Build a tuple from values converted to hashables.
+            out = [_to_hashable(item) for item in obj]
+            return tuple(out)
+
+    elif hasattr(obj, "__hash__"):
```

Comment:
Why do we not want to just do this for iterables? Or in other words, why don't we do this before checking for iterables?

Reply:
Not all iterables are hashable. Iterables which are mutable (e.g. lists) aren't. We have to safely convert them to an immutable first.

Reply:
Why we don't do it first. Consider the following example: `tuple([[1, 2, 3], [4, 5]]). This is a tuple (immutable, hashable), but it contains lists. If we try to call `__hash__` on it, we get the following error: `TypeError: unhashable type: 'list'`.

Reply:
So we first need to iterate through all iterable items, in all levels of depth, and convert everything that's mutable, to an immutable.

Reply:
The function will convert the example I provided to:
```
In[10]: ccl.base._to_hashable(tuple([[1, 2, 3], [4, 5]]))
Out[10]: ((1, 2, 3), (4, 5))
```
which can be safely hashed.

Reply:
but my point is, if whatever the object is has `__hash__`, then we should just use that, correct? If so, why not check that first. Or are you trying to replace the hash of iterables by your own hash?

Reply:
It is normal to overload `__hash__` if the built-in implementation does not cover all cases that the developers need. In our case, the default `__hash__` wouldn't work.

Reply:
ok, that's what I wanted to know